Allow using plugin file names and environment variables compatible with Garden and later

azeey opened the pull request:
# 🎉 New feature

## Summary
Allows using gz-* in plugin filenames (eg.  `<plugin filename="gz-sim-physics-system" name="gz::sim::systems::Physics"/>`) and using `GZ_*` environment variables in Fortress. This is meant to help users (especially ROS users) maintain SDF files and launch files that work in both Fortress and Garden/Harmonic.


## Test it
I've added tests that exercise:
* Using `gz-sim-*` filenames in `<plugin>` tags (`src/SystemLoader_TEST.cc`)
* Using `GZ_SIM_SYSTEM_PLUGIN_PATH` (`src/SystemLoader_TEST.cc`)
* Using `GZ_SIM_RESOURCE_PATH` (`src/Server.cc`)
* Using `GZ_SIM_SERVER_CONFIG_PATH` (`src/ServerConfig_TEST.cc`)

I have not added a test for `GZ_SIM_RENDER_ENGINE_PATH` since I couldn't find a test for `IGN_GAZEBO_RENDER_ENGINE_PATH` and I don't think it's a commonly used environment variable.


## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

mjcarroll commented on include/gz/sim/Util.hh:
```diff
@@ -287,19 +287,31 @@ namespace ignition
+    /// GZ_SIM_RESOURCE_PATH for compatibility with newer versions of Gazebo.
     const std::string kResourcePathEnv{"IGN_GAZEBO_RESOURCE_PATH"};
+    /// \brief Environment variable holding resource paths. 
+    const std::string kResourcePathEnvGzSim{"GZ_SIM_RESOURCE_PATH"};
```
Not actionable here, but we should probably switch these over to `constexpr const char*` at some point.

azeey replied:
Agreed!

iche033 reviewed (commented):
just minor comments otherwise looks good

azeey commented on src/ServerPrivate.hh:
```diff
@@ -122,7 +122,7 @@ namespace ignition
       /// not be found.
       ///
-      /// Fuel will be checked and then the GZ_GAZEBO_RESOURCE_PATH environment
+      /// Fuel will be checked and then the IGN_GAZEBO_RESOURCE_PATH environment
```
Done in https://github.com/gazebosim/gz-sim/pull/2275/commits/68a129a2ddd891d6bab0be1a34eba8fdf8e668b1

azeey commented:
I found that `GZ_SIM_RESOURCE_PATH` wasn't working even though the tests I added were passing. I've made some fixes and added tests in https://github.com/gazebosim/gz-sim/pull/2275/commits/68a129a2ddd891d6bab0be1a34eba8fdf8e668b1. @iche033 PTAL

iche033 commented:
the `UNIT_Server_TEST` failure in `gz_sim_ci-pr_any-focal-amd64` build could be unrelated. The tests pass locally

mjcarroll commented on src/SystemLoader_TEST.cc:
```diff
@@ -30,29 +30,46 @@
 using namespace gz;
 using namespace sim;
 
+#ifdef _WIN32
+  constexpr char *kPluginDir = "bin";
+#else
+  constexpr char *kPluginDir = "lib";
+#endif
```
I think these need to be additionally const-qualified.

```suggestion
#ifdef _WIN32
  constexpr const char *kPluginDir = "bin";
#else
  constexpr const char *kPluginDir = "lib";
#endif
```

mjcarroll replied:
too fast :)

azeey replied:
ha ha, nice!

azeey commented:
CI is good. Windows warnings are preexisting, so I'll go ahead and merge.
